pandas/tests/internals/test_internals.py

```diff
@@ -1101,6 +1134,107 @@ def test_datetime_block_can_hold_element(self):
+        self.check_series_setitem(elem, ii, True)
+        assert blk._can_hold_element(elem)
+
+        # Careful: to get the expected Series-inplace behavior we need
+        # `elem` to not have the same length as `arr`
+        ii2 = IntervalIndex.from_breaks(arr[:-1], closed="neither")
+        elem = element(ii2)
+        self.check_series_setitem(elem, ii, False)
+        assert not blk._can_hold_element(elem)
+
+        ii3 = IntervalIndex.from_breaks([Timestamp(1), Timestamp(3), Timestamp(4)])
+        elem = element(ii3)
+        self.check_series_setitem(elem, ii, False)
+        assert not blk._can_hold_element(elem)
+
+        ii4 = IntervalIndex.from_breaks([Timedelta(1), Timedelta(3), Timedelta(4)])
+        elem = element(ii4)
+        self.check_series_setitem(elem, ii, False)
+        assert not blk._can_hold_element(elem)
+
+    def test_period_can_hold_element_emptylist(self):
+        pi = period_range("2016", periods=3, freq="A")
+        blk = make_block(pi._data, [1], ndim=2)
+
+        assert blk._can_hold_element([])
+
+    def test_period_can_hold_element(self, element):
+        pi = period_range("2016", periods=3, freq="A")
+
+        elem = element(pi)
+        self.check_series_setitem(elem, pi, True)
+
+        # Careful: to get the expected Series-inplace behavior we need
+        # `elem` to not have the same length as `arr`
+        pi2 = pi.asfreq("D")[:-1]
+        elem = element(pi2)
+        self.check_series_setitem(elem, pi, False)
+
+        dti = pi.to_timestamp("S")[:-1]
+        elem = element(dti)
+        self.check_series_setitem(elem, pi, False)
+
+    def check_setting(self, elem, index: Index, inplace: bool):
+        self.check_series_setitem(elem, index, inplace)
+        self.check_frame_setitem(elem, index, inplace)
+
+    def check_can_hold_element(self, obj, elem, inplace: bool):
+        blk = obj._mgr.blocks[0]
+        if inplace:
+            assert blk._can_hold_element(elem)
+        else:
+            assert not blk._can_hold_element(elem)
+
+    def check_series_setitem(self, elem, index: Index, inplace: bool):
+        arr = index._data.copy()
+        ser = Series(arr)
+
+        self.check_can_hold_element(ser, elem, inplace)
+
+        if is_scalar(elem):
+            ser[0] = elem
+        else:
+            ser[: len(elem)] = elem
+
+        if inplace:
+            assert ser.array is arr  # i.e. setting was done inplace
+        else:
+            assert ser.dtype == object
+
+    def check_frame_setitem(self, elem, index: Index, inplace: bool):
+        arr = index._data.copy()
+        df = DataFrame(arr)
+
+        self.check_can_hold_element(df, elem, inplace)
+
+        if is_scalar(elem):
+            df.iloc[0, 0] = elem
+        else:
+            df.iloc[: len(elem), 0] = elem
+
+        if inplace:
+            assert df._mgr.blocks[0].values is arr  # i.e. setting was done inplace
+        else:
+            assert df.dtypes[0] == object
```

Comment:
maybe should think about breaking up this test file at some point

Reply:
ill take a look at some point.  we'll get to delete the test class below in #35417, so thats a start